torchao/experimental/quant_api.py

```diff
@@ -0,0 +1,277 @@
+            * (activation_qvals.reshape(-1, k) - activations_zeros.reshape(-1, 1))
+        ).reshape(m, k)
+        res = torch.matmul(activations_dequantized, weights_dequantized.transpose(1, 0))
+
+        if squeeze_dim0:
+            res = res.unsqueeze(0)
+        return res
+
+
+def _get_quantized_linear_native(nbit, has_weight_zeros):
+    if nbit in [2, 3, 4, 5]:
+        wzp_suffix = "z" if has_weight_zeros else ""
+        return _Int8DynActIntxWeightQuantizedLinearNative(
+            pack_weight_op=getattr(
+                torch.ops.torchao, f"_pack_weights_a8sz_w{nbit}s{wzp_suffix}"
+            ),
+            linear_op=getattr(torch.ops.torchao, f"_linear_a8sz_w{nbit}s{wzp_suffix}"),
+        )
+    raise NotImplementedError(
+        f"Not currently supported: nbit={nbit}, has_weight_zeros={has_weight_zeros}."
+    )
+
+
+def _replace_linear_with_quantized_linear(module: nn.Module, kwargs={}):
+    group_size = kwargs["group_size"]
+    nbit = kwargs["nbit"]
+    has_weight_zeros = kwargs["has_weight_zeros"]
+
+    assert not isinstance(module, nn.Linear)
+    assert nbit >= 1 and nbit <= 7
+
+    for name, child in module.named_children():
+        if not isinstance(child, nn.Linear):
+            _replace_linear_with_quantized_linear(child, kwargs)
+        else:
+            assert child.bias is None
+            qlinear = None
+            try:
+                qlinear = _get_quantized_linear_native(
+                    nbit=nbit,
+                    has_weight_zeros=has_weight_zeros,
+                )
+            except Exception as e:
+                print(
+                    f"Warning: _Int8DynActIntxWeightQuantizedLinearNative raised an exception during initialization: {e}\n"
+                    + "Falling back to **slow** implementation _Int8DynActIntxWeightQuantizedLinearFallback."
+                )
+                qlinear = _Int8DynActIntxWeightQuantizedLinearFallback()
+
+            assert qlinear is not None
+            try:
+                setattr(module, name, qlinear)
+                getattr(module, name).quantize_and_pack_weights(
+                    child.weight, nbit, group_size, has_weight_zeros
+                )
+            except Exception as e:
+                if not isinstance(qlinear, _Int8DynActIntxWeightQuantizedLinearNative):
+                    raise e
+                print(
+                    f"Warning: _Int8DynActIntxWeightQuantizedLinearNative raised an exception during quantize_and_pack_weights: {e}\n"
+                    + "Falling back to **slow** implementation _Int8DynActIntxWeightQuantizedLinearFallback."
+                )
+                qlinear = _Int8DynActIntxWeightQuantizedLinearFallback()
+                setattr(module, name, qlinear)
+                getattr(module, name).quantize_and_pack_weights(
+                    child.weight, nbit, group_size, has_weight_zeros
+                )
+
+
+class Int8DynActIntxWeightQuantizer:
```

Comment:
nit: the file name sounds like an interface but the content looks like all different kinds of implementation....